homeassistant/components/repetier/__init__.py

```diff
@@ -0,0 +1,262 @@
+
+        client = pyrepetier.Repetier(
+            url=url,
+            port=port,
+            apikey=api_key)
+
+        printers = client.getprinters()
+
+        if not printers:
+            return False
+
+        sensors = repetier[CONF_SENSORS][CONF_MONITORED_CONDITIONS]
+        api = PrinterAPI(hass, client, printers, sensors,
+                         repetier[CONF_NAME], config)
+        api.update()
+        track_time_interval(hass, api.update, SCAN_INTERVAL)
+
+        hass.data[REPETIER_API][repetier[CONF_NAME]] = api
+
+    return True
+
+
+class PrinterAPI:
+    """Handle the printer API."""
+
+    def __init__(self, hass, client, printers, sensors, conf_name, config):
+        """Set up instance."""
+        self._hass = hass
+        self._client = client
+        self.printers = printers
+        self.sensors = sensors
+        self.conf_name = conf_name
+        self.config = config
+        self._known_entities = []
+
+    def get_data(self, printer_id, sensor_type, temp_id):
+        """Get data from the state cache."""
+        printer = self.printers[printer_id]
+        methods = API_PRINTER_METHODS[sensor_type]
+        for prop, offline in methods['offline'].items():
+            state = getattr(printer, prop)
+            if state == offline:
+                # if state matches offline, sensor is offline
+                return None
+
+        data = {}
+        for prop, attr in methods['state'].items():
+            prop_data = getattr(printer, prop)
+            if attr == 'temp_data':
+                temp_methods = methods['temp_data']
+                for temp_prop, temp_attr in temp_methods.items():
+                    data[temp_attr] = getattr(prop_data[temp_id], temp_prop)
+            else:
+                data[attr] = prop_data
+        return data
+
+    def update(self, now=None):
+        """Update the state cache from the printer API."""
+        self._load_entities()
+        for printer in self.printers:
+            printer.get_data()
+        dispatcher_send(self._hass, UPDATE_SIGNAL)
+
+    def _load_entities(self):
+        sensor_info = []
+        for pidx, printer in enumerate(self.printers):
+            for sensor_type in self.sensors:
+                info = {}
+                info['sensor_type'] = sensor_type
+                info['printer_id'] = pidx
+                info['name'] = printer.slug
+                info['printer_name'] = self.conf_name
+                known = '{}-{}'.format(printer.slug, sensor_type)
+
+                if sensor_type == 'bed_temperature':
+                    if printer.heatedbeds is None:
+                        continue
+                    if any(known in s for s in self._known_entities):
+                        continue
+                    for idx, _ in enumerate(printer.heatedbeds):
+                        info['temp_id'] = idx
+                        sensor_info.append(info)
+                    self._known_entities.append(known)
```

Comment:
This can be moved below the sensor type checks since it's done for all sensor type cases.

Reply:
Brain fart from my side :D - of course no need to call it in every check :)